Notify if qube tried to send large clipboard

alimirjamali opened the pull request:
fixes: https://github.com/QubesOS/qubes-issues/issues/9296

marmarta commented on qui/clipboard.py:
```diff
@@ -75,40 +77,110 @@ def my_init(self, loop=None, gtk_app=None):
+            body = _("Malformed clipboard request received from qube: "
+                     "<b>{vmname}</b>!").format(vmname=metadata["vmname"])
+        elif metadata["qrexec_clipboard"] and \
+                metadata["sent_size"] >= metadata["buffer_size"]:
```
Why is this not treated as oversized request? As in, why is this a separate elif?

alimirjamali replied:
Because Microsoft Windows uses another path to send clipboard to global clipboard (qrexec instead of X11 gui daemon). And the patch to gui-daemon and gui-agent was getting extremely complex and big. Developing a mechanism for Microsoft Windows to notify oversized clipboards to GUIVM needs additional effort. So the GUIVM has no mechanism to detect of to detect Windows oversized requests.

I guess it would be OK to work on this again after official support for Microsoft Windows is restored.

marmarta reviewed (changes requested):
I really like this! One thing is that I feel like now we have a bunch of different notifications with very different wording, and it would be good to have them more unified. A proposal:

1. warning and errors probably should have error icon, see disk space widget for example icons; I think we can use dialog-warning right now for everything and I am working on more unified icon experience

* malformed data
<b>ERROR</b>: Malformed clipboard data received from qube: <b>{vmname}</b>

* error on copy
<b>ERROR</b>: Failed to fetch clipboard data from qube: <b>{vmname}</b>

* error on paste
ERROR: Failed to paste global clipboard contents to qube: <b>{vmname}</b>

* oversize data
<b>ERROR</b>: Global clipboard size exceeded.\nQube <b>{vmname}</b> attempted to send {size} bytes to global clipboard.\nCurrent global clipboard limit is {limit}, increase limit or use <i>qvm-copy</i> to transfer large amounts of data between qubes. 

* oversize data, possible truncation
<b>WARNING</b>: Global clipboard size limit exceed.\nQube <b>{vmname}</b>  attempted to send {size} bytes to global clipboard.\nGlobal clipboard might have been truncated.
<small>Use <i>qvm-copy</i> to transfer large amounts of data between qubes.</small>

* empty clipboard
<b>WARNING</b>: Empty qube clipboard.\nQube <b>{vmname}</b> attempted to send <b>0</b> bytes to global clipboard.

2. the following don't need an  icon
* success

Clipboard contents fetched from qube: <b>'{vmname}'</b>\n
Copied <b>{size}</b> to the global clipboard.\n
<small>Press {shortcut} in qube to paste to local clipboard.</small>

Global clipboard copied <b>{size}</b> to <b>{vmname}</b>. Global clipboard has been wiped.\n<small>Paste normally in qube (e.g. Ctrl+V).</small>".

Global clipboard copied to qube and wiped.<i/>\n<small>Paste normally in qube (e.g. Ctrl+V).</small>

* additional message
\n<small>Global clipboard has been wiped</small>

I'm very open to corrections here, just trying for more unified messages and make it more explicit that an error is an error.

alimirjamali commented:
> I really like this! One thing is that I feel like now we have a bunch of different notifications with very different wording, and it would be good to have them more unified.

Excellent suggestions. I hope that the tests & review of patches to `qubes-gui-daemon`, `qubes-gui-agent` and `qubes-gui-common` would be successful and they are merged. Since they are the backbone to this PR and they provide the .metadata file.

There is more work to be done. The patches to global clipboard will allow different limits per VM (or per entire GUIVM). Using specific features which were explained in the original issue comments. And there should be a place (maybe on Global Config) to allow users to easily set them.

In addition to that, I am working on [issue #6641](https://github.com/QubesOS/qubes-issues/issues/6641) (option to Clear inter-VM global clipboard after a certain amount of time). That one also should be added to Global Config. Overall if all of these are implemented, Qubes OS 4.3 clipboard could get much better and smarter.

alimirjamali commented:
> 1. warning and errors probably should have error icon, see disk space widget for example icons; I think we can use dialog-warning right now for everything and I am working on more unified icon experience

I added the support for the icons and moved all messages to top of the file. So it would be easier for review & modifications. Also applied suggestions. I think some messages might need additional work. Some of the lines are too long.

In the end, we should still wait for the gui-daemon and gui-agent patches to be merged for all of this to work.


https://github.com/user-attachments/assets/7101a7ab-b90c-4bfd-b622-904455f1648f

marmarta commented:
> There is more work to be done. The patches to global clipboard will allow different limits per VM (or per entire GUIVM). Using specific features which were explained in the original issue comments. And there should be a place (maybe on Global Config) to allow users to easily set them.
> 

Yeah, I think in the future Global Config -> Clipboard will be the place for this.
